internals/overlord/logstate/loki/loki.go

```diff
@@ -0,0 +1,211 @@
+	"net/http"
+	"sort"
+	"strconv"
+	"strings"
+	"time"
+
+	"github.com/canonical/pebble/internals/overlord/logstate/clienterr"
+	"github.com/canonical/pebble/internals/plan"
+	"github.com/canonical/pebble/internals/servicelog"
+)
+
+const maxRequestSize = 1000
+
+var requestTimeout = 10 * time.Second
+
+type Client struct {
+	target    *plan.LogTarget
+	remoteURL string
+	// buffered entries are "sharded" by service name
+	entries    map[string][]lokiEntry
+	numEntries int
+
+	httpClient http.Client
+	retryAfter *time.Time
+}
+
+func NewClient(target *plan.LogTarget) *Client {
+	return &Client{
+		target:     target,
+		remoteURL:  target.Location,
+		entries:    map[string][]lokiEntry{},
+		httpClient: http.Client{Timeout: requestTimeout},
+	}
+}
+
+func (c *Client) Write(ctx context.Context, entry servicelog.Entry) error {
+	c.entries[entry.Service] = append(c.entries[entry.Service], encodeEntry(entry))
+	c.numEntries++
+
+	if c.numEntries >= maxRequestSize {
+		if c.retryAfter != nil {
+			if time.Now().Compare(*c.retryAfter) == -1 {
+				// Retry-After deadline hasn't passed yet, so we shouldn't flush
+				return nil
+			}
+			c.retryAfter = nil
+		}
+		return c.Flush(ctx)
+	}
+	return nil
+}
+
+func encodeEntry(entry servicelog.Entry) lokiEntry {
+	return lokiEntry{
+		strconv.FormatInt(entry.Time.UnixNano(), 10),
+		strings.TrimSuffix(entry.Message, "\n"),
+	}
+}
+
+func (c *Client) Flush(ctx context.Context) error {
+	if c.numEntries == 0 {
+		return nil // no-op
+	}
+	defer c.emptyBuffer()
+
+	req := c.buildRequest()
+	jsonReq, err := json.Marshal(req)
+	if err != nil {
+		return fmt.Errorf("encoding request to JSON: %v", err)
+	}
+
+	httpReq, err := http.NewRequestWithContext(ctx, http.MethodPost, c.remoteURL, bytes.NewReader(jsonReq))
+	if err != nil {
+		return fmt.Errorf("creating HTTP request: %v", err)
+	}
+	httpReq.Header.Set("Content-Type", "application/json")
+
+	resp, err := c.httpClient.Do(httpReq)
+	if err != nil {
+		return handleHTTPClientErr(err)
+	}
+	return c.handleServerResponse(resp)
+}
+
+func (c *Client) emptyBuffer() {
+	for svc := range c.entries {
+		c.entries[svc] = c.entries[svc][:0]
+	}
+	c.numEntries = 0
+}
+
+func (c *Client) buildRequest() request {
+	// Sort keys to guarantee deterministic output
+	var services []string
+	for svc, entries := range c.entries {
+		if len(entries) == 0 {
+			continue
+		}
+		services = append(services, svc)
+	}
+	sort.Strings(services)
+
+	var req request
+	for _, service := range services {
+		entries := c.entries[service]
+		stream := stream{
+			Labels: map[string]string{
+				"pebble_service": service,
+			},
+			Entries: entries,
+		}
+		req.Streams = append(req.Streams, stream)
+	}
+	return req
+}
+
+type request struct {
+	Streams []stream `json:"streams"`
+}
+
+type stream struct {
+	Labels  map[string]string `json:"stream"`
+	Entries []lokiEntry       `json:"values"`
+}
+
+type lokiEntry [2]string
+
+func handleHTTPClientErr(err error) error {
+	var netErr net.Error
```

Comment:
Related to discussion elsewhere: this is one of the few cases where an error is not a pointer, because `net.Error` is defined as an interface.